mod_papers/helper.php

```diff
@@ -13,150 +14,257 @@
+                        usleep(3000);
+		} while ($mrc == CURLM_CALL_MULTI_PERFORM);
+
+		while ($active && $mrc == CURLM_OK) {
+			// Wait for activity on any curl-connection
+			if (curl_multi_select($mh) == -1) {
+                        	usleep(3000);
+			}
+				// Continue to exec until curl is ready to
+				// give us more data
+				do {
+					$mrc = curl_multi_exec($mh, $active);
+				} while ($mrc == CURLM_CALL_MULTI_PERFORM);
+			
+		}
+
+		// get content and remove handles
+		$mergedworks = array();
+		foreach($multiCurl as $k => $ch) {
+			$mdata[$k] = json_decode(curl_multi_getcontent($ch), true);
+			curl_multi_remove_handle($mh, $ch);
+			$works = $mdata[$k]['group'];
+			if (!empty($works)) {
+				$mergedworks = array_merge($mergedworks, $works);
+			}
+		}
+		// close
+		curl_multi_close($mh);
+
+		//Get all dois
+		$dois = array();
+		foreach ($mergedworks as $key => $work) {
+			if (!is_null($work['external-ids'])) {
+
+				foreach ($work['external-ids']['external-id'] as $ids) {
+					if (strcmp($ids['external-id-type'], 'doi') == 0) {
+						$dois[] = $ids['external-id-value'];
+					}
+				}
+				$mergedworks[$key] = array_merge($mergedworks[$key], array('parse'=>1,'udoi'=>'')); //Build a parse check field. Parse by default, set to zero if there's an issue.
+			} else {
+				unset($mergedworks[$key]); //For now, kill anything without a DOI.
+			}
+		}
+
+		//Find all unique dois
+		$udois = array_unique($dois);
+
+		//sanitise merged array.
+		foreach ($mergedworks as $mkey => $work) {
+			//Identify Duplicates
+			foreach ($work['external-ids']['external-id'] as $ids) {
+				if (strcmp($ids['external-id-type'], 'doi') == 0) {
+					$doi = $ids['external-id-value'];
+					$mergedworks[$mkey]['udoi'] = $doi;
+					$key = array_search($doi, $udois); // Find where DOI is in the unique list
+
+					unset($udois[$key]); //Found one, don't need another.
+					if ($key === false) {
+						$mergedworks[$mkey]['parse'] = 0; //Don't parse this entry
+					}
+				}
+			}
+		}
+
+		//Sort array by year
+		usort($mergedworks, function($a, $b) {
+			return ($a['work-summary']['0']['publication-date']['year']['value'] > $b['work-summary']['0']['publication-date']['year']['value']) ? -1 : 1;
+		});
+
+		$curr_year = date("Y");
+		$output = "<h2>" . $curr_year . "</h2>";
+
+		// array of curl handles
+		$multiCurl = array();
+		// data to be returned
+		$mdata = array();
+
+		$mh  = curl_multi_init();
+		$id = 0;
+		foreach ($mergedworks as $work) {
+			//Identify Results earlier than 2011
+			$year = $work['work-summary']['0']['publication-date']['year']['value'];
+			if ($year < '2011') {
+				$work['parse'] = 0; //Don't parse this entry
+			}
+			//Print results
+			if ($work['parse'] === 1) {
+
+				// create a new cURL resource
+				// set URL and other appropriate options
+				$ch  = curl_init();
+				$options = array(
+					CURLOPT_URL => 'https://pub.orcid.org/v2.1' . $work['work-summary']['0']['path'],
```

Comment:
This part here is where I see a problem. Before you put in the multi curl portion this was taking around 35 seconds to complete. Not sure how much it is now due to your cache system (as in each time I request the site, 99 out of 100 times I'll get the cached version).

Regardless, this is going to call the API `count($work)` times. If the list gets long enough the OrcID servers are going to rate limit us and it doesn't matter how asynchronous our calls are.

In #2 I suggested calling `/v2.1/{orcid}/works/{putCodes}` here, then we'd only have three new API calls rather than dosens. Did you attempt that already and hit a roadblock? Or is this something worth trying?

Reply:
I will try to refactor it

Reply:
If I understood correctly, you propose to call /v2.1/{orcid}/works/{putCode1, putCode2, putCode3....}
I have tried this and has faced number of putCodes limit - it allows only 50 items maximum and I had more records per one orcid ID.  How could overcome this? Should I make batches of queries? Is it that such syntax is equivalent to curl_multi call?